feat(vector db create form): updated to add description , tag and metadata (#469)

SubhajitMohanty1 opened the pull request:
## Description
Implemented backend support for description , tag and metadata fields to the Vector DB create form.

## Changes Made
We made changes to the AbstractVectorDatabaseEngine.java , PGVectorDatabaseEngine.java , SecurityEngineUtils.java , VectorDatabaseParamOptionsEnum.java , CreateVectorDatabaseEngineReactor.java and Constants.java

SubhajitMohanty1 committed:
feat(vector db create form): updated to add description , tag and metadata (#469)

CodiumAI-Agent commented:
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/SEMOSS/Semoss/pull/877/files#diff-6924060b7da9e1d7e54c0dd4ddb64220e24f36a3e6519ac2a00106fb928bf859R426-R435'><strong>SplitTextCommand Bug</strong></a>

The order of appended parameters in splitTextCommand is incorrect, causing wrong values for chunk_overlap, description, tags, and metaData in the generated command.
</summary>

```java
.append("', chunk_size = ")
.append(chunkMaxTokenLength)
.append(", chunk_overlap = ")
.append(engineDescription)
.append(", description = ")
.append(engineTags)
.append(", tags = ")
.append(engineMetaData)
.append(", metaData = ")
.append(tokenOverlapBetweenChunks)
```

</details>

<details><summary><a href='https://github.com/SEMOSS/Semoss/pull/877/files#diff-066bbdc0dbfade3fc85cd17585a6c61c9be52a2a027467818a86cc9d696c3e5eR1084-R1092'><strong>SplitTextCommand Bug</strong></a>

The new append calls misplace engineDescription, engineTags, engineMetaData, and tokenOverlapBetweenChunks, resulting in an invalid split_text invocation.
</summary>

```java
.append(chunkMaxTokenLength)
.append(", chunk_overlap = ")
.append(engineDescription)
.append(", description = ")
.append(engineTags)
.append(", tags = ")
.append(engineMetaData)
.append(", metaData = ")
.append(tokenOverlapBetweenChunks)
```

</details>

<details><summary><a href='https://github.com/SEMOSS/Semoss/pull/877/files#diff-499e7f2fe87d9e3b63c3988b707030389148db2a98164be3ff2a03b36eaba1b3R170-R178'><strong>Missing Metadata Persistence</strong></a>

The metadata field is not persisted via SecurityEngineUtils.addTags, and the implementation uses hardcoded metaKey strings instead of existing Constants.
</summary>

```java
if(vectorDbDetails.get(Constants.TAGS)!=null && !vectorDbDetails.get(Constants.TAGS).toString().equals("")) {
	SecurityEngineUtils.addTags(vectorDbId,"tag",vectorDbDetails.get(Constants.TAGS).toString());
}

if(vectorDbDetails.get(Constants.DESCR)!=null && !vectorDbDetails.get(Constants.DESCR).toString().equals("")) {
	SecurityEngineUtils.addTags(vectorDbId,"description",vectorDbDetails.get(Constants.DESCR).toString());
}


```

</details>

</td></tr>
</table>

CodiumAI-Agent commented:
## PR Code Suggestions ✨

<!-- 6f3b618 -->

<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Correct parameter ordering and quoting</summary>

___


**The parameters for <code>chunk_overlap</code>, <code>description</code>, <code>tags</code>, and <code>metaData</code> are in the wrong <br>order and missing string quoting. Swap <code>tokenOverlapBetweenChunks</code> to the <br><code>chunk_overlap</code> slot and wrap string values in quotes.**

[src/prerna/engine/impl/vector/AbstractVectorDatabaseEngine.java [428-435]](https://github.com/SEMOSS/Semoss/pull/877/files#diff-6924060b7da9e1d7e54c0dd4ddb64220e24f36a3e6519ac2a00106fb928bf859R428-R435)

```diff
 splitTextCommand.append(", chunk_overlap = ")
+    .append(tokenOverlapBetweenChunks)
+    .append(", description = '")
     .append(engineDescription)
-    .append(", description = ")
+    .append("'")
+    .append(", tags = '")
     .append(engineTags)
-    .append(", tags = ")
+    .append("'")
+    .append(", metaData = '")
     .append(engineMetaData)
-    .append(", metaData = ")
-    .append(tokenOverlapBetweenChunks)
+    .append("'");
```
<details><summary>Suggestion importance[1-10]: 9</summary>

__

Why: The code incorrectly assigns `engineDescription` to `chunk_overlap`, causing wrong parameter ordering and missing quotes, which is a critical bug in the command.


</details></details></td><td align=center>High

</td></tr><tr><td>



<details><summary>Fix PGVector splitTextCommand</summary>

___


**Just like in the abstract engine, the argument order is incorrect and string values <br>need quoting. Place <code>tokenOverlapBetweenChunks</code> first and wrap each metadata string in <br>single quotes.**

[src/prerna/engine/impl/vector/PGVectorDatabaseEngine.java [1085-1092]](https://github.com/SEMOSS/Semoss/pull/877/files#diff-066bbdc0dbfade3fc85cd17585a6c61c9be52a2a027467818a86cc9d696c3e5eR1085-R1092)

```diff
 splitTextCommand.append(", chunk_overlap = ")
+    .append(tokenOverlapBetweenChunks)
+    .append(", description = '")
     .append(engineDescription)
-    .append(", description = ")
+    .append("'")
+    .append(", tags = '")
     .append(engineTags)
-    .append(", tags = ")
+    .append("'")
+    .append(", metaData = '")
     .append(engineMetaData)
-    .append(", metaData = ")
-    .append(tokenOverlapBetweenChunks)
+    .append("'");
```
<details><summary>Suggestion importance[1-10]: 9</summary>

__

Why: Similar to the abstract engine, the metadata order is wrong and lacks quotes, leading to incorrect command syntax, so fixing is crucial.


</details></details></td><td align=center>High

</td></tr><tr><td rowspan=2>General</td>
<td>



<details><summary>Use try-with-resources for statements</summary>

___


**Use try-with-resources for <code>PreparedStatement</code> to ensure it’s closed properly and <br>switch to <code>executeUpdate()</code> for clarity. Also, minimize repeated calls to <br><code>ps.getConnection()</code>.**

[src/prerna/auth/utils/SecurityEngineUtils.java [3184-3205]](https://github.com/SEMOSS/Semoss/pull/877/files#diff-44b3fcc23dd64eb7cccf7857fbebd6ca9626a23a3334aa2cacc850a49884f64eR3184-R3205)

```diff
 public static void addTags(String engineId, String metaKey, String metaValue) {
     String query = "INSERT INTO ENGINEMETA (ENGINEID, METAKEY, METAVALUE, METAORDER) VALUES (?,?,?,?)";
-
-    PreparedStatement ps = null;
-    try {
-        ps = securityDb.getPreparedStatement(query);
-        int parameterIndex = 1;
-        ps.setString(parameterIndex++, engineId);
-        ps.setString(parameterIndex++, metaKey);
-        ps.setString(parameterIndex++, metaValue);
-        ps.setInt(parameterIndex++, 0);
-        ps.execute();
-        if(!ps.getConnection().getAutoCommit()) {
-            ps.getConnection().commit();
+    try (PreparedStatement ps = securityDb.getPreparedStatement(query)) {
+        int idx = 1;
+        ps.setString(idx++, engineId);
+        ps.setString(idx++, metaKey);
+        ps.setString(idx++, metaValue);
+        ps.setInt(idx, 0);
+        ps.executeUpdate();
+        Connection conn = ps.getConnection();
+        if (!conn.getAutoCommit()) {
+            conn.commit();
         }
     } catch (SQLException e) {
         classLogger.error(Constants.STACKTRACE, e);
-    } finally {
-        ConnectionUtils.closeAllConnectionsIfPooling(securityDb, ps);
     }
 }
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Using try-with-resources and `executeUpdate()` improves resource management and clarity and prevents potential leaks.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Use constants and include metadata</summary>

___


**Replace hardcoded metaKey strings with the corresponding <code>Constants</code> fields and add a <br>clause for metadata. This ensures consistency and includes the new metadata field.**

[src/prerna/reactor/vector/upload/CreateVectorDatabaseEngineReactor.java [170-176]](https://github.com/SEMOSS/Semoss/pull/877/files#diff-499e7f2fe87d9e3b63c3988b707030389148db2a98164be3ff2a03b36eaba1b3R170-R176)

```diff
-if(vectorDbDetails.get(Constants.TAGS)!=null && !vectorDbDetails.get(Constants.TAGS).toString().equals("")) {
-    SecurityEngineUtils.addTags(vectorDbId,"tag",vectorDbDetails.get(Constants.TAGS).toString());
+if (vectorDbDetails.get(Constants.TAGS) != null && !vectorDbDetails.get(Constants.TAGS).toString().isEmpty()) {
+    SecurityEngineUtils.addTags(vectorDbId, Constants.TAGS, vectorDbDetails.get(Constants.TAGS).toString());
+}
+if (vectorDbDetails.get(Constants.DESCR) != null && !vectorDbDetails.get(Constants.DESCR).toString().isEmpty()) {
+    SecurityEngineUtils.addTags(vectorDbId, Constants.DESCR, vectorDbDetails.get(Constants.DESCR).toString());
+}
+if (vectorDbDetails.get(Constants.METADATA) != null && !vectorDbDetails.get(Constants.METADATA).toString().isEmpty()) {
+    SecurityEngineUtils.addTags(vectorDbId, Constants.METADATA, vectorDbDetails.get(Constants.METADATA).toString());
 }
 
-if(vectorDbDetails.get(Constants.DESCR)!=null && !vectorDbDetails.get(Constants.DESCR).toString().equals("")) {
-    SecurityEngineUtils.addTags(vectorDbId,"description",vectorDbDetails.get(Constants.DESCR).toString());
-}
-
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: Replacing hardcoded keys with `Constants` enhances consistency and adding `METADATA` covers the new field, improving maintainability.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

SubhajitMohanty1 committed:
feat(vector db create form): latest updates to add description , tag and metadata (#469)

Pragya7011 commented on src/prerna/reactor/vector/upload/CreateVectorDatabaseEngineReactor.java:
@SubhajitMohanty1  : Why metadata is not added as valid check ?Add metadata also in metaValues .

SubhajitMohanty1 committed:
feat(vector db create form): new updates to add description , tag and metadata (#469)

SubhajitMohanty1 committed:
Merge branch 'dev' of https://github.com/SEMOSS/Semoss into 469-AddFieldsToVectorDBForm

SubhajitMohanty1 committed:
feat(vector db create form): updated (#469)

SubhajitMohanty1 commented:
@themaherkhalil : As per UploadEngineReactor, we have updateEngineMetadata method for updating the enginemeta table. So we are now using the same method to update the metavalues for the vector engine as well. We have updated the code changes in SecurityEngineUtils as per your review comment and added new code changes in CreateVectorDatabaseEngineReactor.
Below are the screenshots of both the reactors.
![Screenshot 2025-06-10 171423](https://github.com/user-attachments/assets/9272d070-4761-438e-b707-89769482495b)
![Screenshot 2025-06-10 171608](https://github.com/user-attachments/assets/9b1dcaae-d733-493a-837b-f3df50056e2b)
Kindly Review the code changes.
